### PR TITLE
feat: Context management for using `with` that manages trace(s) automatically

### LINKED ISSUE #1036: investigate ways to clean sdk
with a myriad of integrations now, we should be in a position to identify patterns and abstractions between integrations

lets see how we can simplify and clean up the sdk to make it more maintainable and extensible for the future :)

### PR DESCRIPTION
**Pull request by dot-agi**
## 📥 Pull Request

**📘 Description**

Added native context manager support to TraceContext class, enabling `with agentops.start_trace("name") as trace:` syntax for automatic trace lifecycle management.

**Basic Usage:**
```python
import agentops

agentops.init(api_key="your-api-key")

# Automatic trace management
with agentops.start_trace("my_workflow") as trace:
    # Your code here
    pass  # Trace automatically ends with Success/Error status

# With tags
with agentops.start_trace("data_processing", tags=["batch", "production"]):
    process_data()
```

**Changelog:**
- Added `__enter__` and `__exit__` methods to `TraceContext` class in `agentops/sdk/core.py`
- Automatic trace state management: Success on normal exit, Error on exceptions
- Support for parallel traces in nested context managers (independent, not parent-child)
- Thread-safe and asyncio-compatible implementation
- Added 4 example files: basic usage, parallel traces, error handling, production patterns
- Created comprehensive documentation at `docs/v2/usage/context-managers.mdx`
- Updated SDK reference with context manager usage patterns
- Maintains full backward compatibility with existing manual trace management

**🧪 Testing**

- Context manager protocol tests (`__enter__`, `__exit__` behavior)
- Exception handling and automatic state management tests
- Thread safety and asyncio compatibility tests
- System exception handling tests (KeyboardInterrupt, SystemExit)
- Generator function integration and early return tests
- Large data handling and resource cleanup failure tests
- Backward compatibility and mixed usage pattern tests
- Performance tests with 100+ sequential traces
- Edge case tests (None returns, initialization failures)

### CODE REVIEW

File: agentops/sdk/core.py [hunk 1]
```diff
@@ -34,6 +34,28 @@ def __init__(self, span: Span, token: Optional[context_api.Token] = None, is_ini
         self.span = span
         self.token = token
         self.is_init_trace = is_init_trace  # Flag to identify the auto-started trace
+        self._end_state = "Indeterminate"  # Default end state because we don't know yet
```
[the remark this answers is not in the document]
Comment (dot-agi) [reply agentops/sdk/core.py]: Done using OTel's `StatusCode` codes.

File: agentops/__init__.py [hunk 2]
```diff
@@ -27,6 +27,21 @@ def get_client() -> Client:
     """Get the singleton client instance"""
     global _client
 
+    # If _client is None create a new one
+    if _client is None:
```
[the remark this answers is not in the document]
Comment (dot-agi) [reply agentops/__init__.py:35]: Done.

File: agentops/__init__.py [hunk 3]
```diff
@@ -27,6 +27,21 @@ def get_client() -> Client:
...
+
+    # Check if the current client instance is still valid
+    # This handles the case where test fixtures reset the Client singleton
+    # Only check if we have a real Client instance (not a mock)
+    if hasattr(_client, "__class__") and _client.__class__.__name__ == "Client":
+        # The Client class uses __instance as its singleton, so we check if it's changed
+        current_singleton = Client()
+
+        # If the singleton has changed (due to test resets), update our reference
+        if _client is not current_singleton:
```
[the remark this answers is not in the document]
Comment (dot-agi) [reply agentops/__init__.py]: Done.

File: agentops/__init__.py [hunk 4]
```diff
@@ -15,18 +15,28 @@
...
+_client = None
 
 
 def get_client() -> Client:
-    """Get the singleton client instance"""
+    """Get the singleton client instance in a thread-safe manner"""
     global _client
 
+    # Double-checked locking pattern for thread safety
+    if _client is None:
```
[the remark this answers is not in the document]
Comment (dot-agi) [reply agentops/__init__.py:35]: Cool

File: agentops/sdk/core.py [hunk 5]
```diff
@@ -36,6 +37,38 @@ def __init__(self, span: Span, token: Optional[context_api.Token] = None, is_ini
         self.span = span
         self.token = token
         self.is_init_trace = is_init_trace  # Flag to identify the auto-started trace
+        self._end_state = StatusCode.UNSET  # Default end state because we don't know yet
```
[the remark this answers is not in the document]
Comment (dot-agi) [reply agentops/sdk/core.py:40]: Ok.

File: agentops/enums.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
+"""
```
Comment (tcdent) [inline agentops/enums.py:1]: I don't love files named `types` or `enums` cuz they don't really tell you anything, but if you can't find a module thats directly related to this, it's fine to put them here for now.

File: agentops/sdk/README.md [hunk 7]
```diff
@@ -176,11 +176,11 @@ flowchart TD
...
 ```python
-from agentops import Session, agent, tool
-from agentops.sdk import TracingCore, TracingConfig
+from agentops import Session, agent, tool, tracer
+from agentops.sdk import TracingConfig
 
-# Initialize the tracing core with a dedicated configuration
-TracingCore.get_instance().initialize(
+# Initialize the global tracer with a dedicated configuration
+tracer.initialize(
```
Comment (tcdent) [inline agentops/sdk/README.md:183]: Good catch!

File: agentops/__init__.py [hunk 8]
```diff
@@ -207,7 +225,9 @@ def start_trace(
     return tracer.start_trace(trace_name=trace_name, tags=tags)
```
Comment (tcdent) [inline agentops/__init__.py:225]: So using `with start_trace(...)` just works because otel already returns a context manager?

**Review by tcdent — CHANGES_REQUESTED** [review]

File: agentops/enums.py (hunk 6 above)
Comment (dot-agi) [reply agentops/enums.py:1]: I agree, namespace conflicts are a real PITA. I created this file for a lack of better structure but I have seen no recent module named `enums`.

File: agentops/sdk/README.md (hunk 7 above)
Comment (dot-agi) [reply agentops/sdk/README.md:183]: Thanks!

File: agentops/__init__.py (hunk 8 above)
Comment (dot-agi) [reply agentops/__init__.py:225]: Not OTel directly but `agentops` returns a `TraceContext` that handles context management for us.

**Review by tcdent — APPROVED** [review]
